airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3Destination.java

```diff
@@ -106,42 +105,69 @@ public static void testMultipartUpload(final AmazonS3 s3Client, final String buc
       }
+      success = true;
+    } finally {
+      if (success) {
+        manager.complete();
+      } else {
+        manager.abort();
+      }
+      s3Client.deleteObject(bucketName, testFile);
     }
-
-    manager.complete();
-    s3Client.deleteObject(bucketName, testFile);
-
     LOGGER.info("Finished verification for multipart upload mode");
   }
 
-  @Override
-  public AirbyteMessageConsumer getConsumer(final JsonNode config,
-                                            final ConfiguredAirbyteCatalog configuredCatalog,
-                                            final Consumer<AirbyteMessage> outputRecordCollector) {
-    final S3WriterFactory formatterFactory = new ProductionWriterFactory();
-    return new S3Consumer(S3DestinationConfig.getS3DestinationConfig(config), configuredCatalog, formatterFactory, outputRecordCollector);
-  }
-
   /**
    * Note that this method completely ignores s3Config.getBucketPath(), in favor of the bucketPath
    * parameter.
    */
-  public static void attemptS3WriteAndDelete(final S3DestinationConfig s3Config, final String bucketPath) {
-    attemptS3WriteAndDelete(s3Config, bucketPath, s3Config.getS3Client());
+  public static void attemptS3WriteAndDelete(final S3StorageOperations storageOperations,
+                                             final S3DestinationConfig s3Config,
+                                             final String bucketPath) {
+    attemptS3WriteAndDelete(storageOperations, s3Config, bucketPath, s3Config.getS3Client());
   }
 
   @VisibleForTesting
-  static void attemptS3WriteAndDelete(final S3DestinationConfig s3Config, final String bucketPath, final AmazonS3 s3) {
+  static void attemptS3WriteAndDelete(final S3StorageOperations storageOperations,
+                                      final S3DestinationConfig s3Config,
+                                      final String bucketPath,
+                                      final AmazonS3 s3) {
     final var prefix = bucketPath.isEmpty() ? "" : bucketPath + (bucketPath.endsWith("/") ? "" : "/");
     final String outputTableName = prefix + "_airbyte_connection_test_" + UUID.randomUUID().toString().replaceAll("-", "");
-    attemptWriteAndDeleteS3Object(s3Config, outputTableName, s3);
+    attemptWriteAndDeleteS3Object(storageOperations, s3Config, outputTableName, s3);
   }
 
-  private static void attemptWriteAndDeleteS3Object(final S3DestinationConfig s3Config, final String outputTableName, final AmazonS3 s3) {
+  private static void attemptWriteAndDeleteS3Object(final S3StorageOperations storageOperations,
+                                                    final S3DestinationConfig s3Config,
+                                                    final String outputTableName,
+                                                    final AmazonS3 s3) {
     final var s3Bucket = s3Config.getBucketName();
 
+    storageOperations.createBucketObjectIfNotExists(s3Bucket);
```

Comment:
the permission check fails if the bucket does not exist yet.
So, i added the bucket creation before running the read/write/list tests

Reply:
Destinations usually go on and create stage areas, schemas, datasets, tables if they don't exist and you need to have one for the connector to do its work. 

In the Airbyte UI, the user can also come up with some new names for these objects, and the UX would be more difficult if you have to go to another tool to create them before entering one that already exists in Airbyte.

Reply:
ah yeah, makes sense 👍